ion-schema/src/constraint.rs

```diff
@@ -731,72 +733,97 @@ impl OrderedElementsConstraint {
+    // +------------------------------+
+    // | event  | State Visits        |
+    // +------------------------------+
+    // |   -    |  I: 1               |
+    // |   1    |  S1: 1, S2: 1       |
+    // |   2    |  S2: 1, S3: 1       |
+    // |   END  |  S3: 1              |
+    // +------------------------------+
+    // As shown above visit count for `END` doesn't have final state in it which means the value resulted to be invalid.
+    //
+    fn build_nfa_from_type_ids(type_ids: &Vec<TypeId>, type_store: &TypeStore) -> NfaRun {
+        let mut nfa_builder = NfaBuilder::new();
+        // add initial state to nfa
+        nfa_builder.with_state(State::Initial);
+
+        for type_id in type_ids {
+            let type_def = type_store.get_type_by_id(*type_id).unwrap();
+            let occurs_range: Range = type_def.get_occurs_constraint("ordered_elements");
+
+            // unwrap here won't lead to panic as the check for non negative range was already done while parsing ordered_elements constraint
+            let (min, max) = occurs_range.non_negative_range_boundaries().unwrap();
+
+            // add transition to next state
+            nfa_builder.with_transition(
+                nfa_builder.total_states() - 1,
+                nfa_builder.total_states(),
+                Some(*type_id),
+                min == 0,
+            );
+
+            nfa_builder.with_state(State::Intermediate { min, max });
+
+            if max >= 2 {
+                // add a a transition to self for states that have  max >= 2
+                nfa_builder.with_transition(
+                    nfa_builder.total_states() - 1,
+                    nfa_builder.total_states() - 1,
+                    Some(*type_id),
+                    min == 0,
+                );
             }
-
-            ion_path.pop();
-            index += 1;
         }
 
-        // greedily take as many values as we can of this type without going out
-        // of the maximum of the range
-        while Option::is_some(&values_iter.peek()) && occurs_range.contains(&(count + 1).into()) {
-            // don't consume it until we know it's valid for the type
-            if let Some(value) = values_iter.peek() {
-                let schema_element: IonSchemaElement = (*value).into();
-                if type_def.is_valid(&schema_element, type_store, ion_path) {
-                    let _ = values_iter.next(); // consume it as it is valid
-                    count += 1;
-                } else {
-                    // if the value doesn't match this type_def, then we'll break out of the while
-                    // loop and check the value against the next type_def.
-                    break;
-                }
-            }
-        }
+        // add final state transition
+        nfa_builder.with_transition(
+            nfa_builder.total_states() - 1,
+            nfa_builder.total_states(),
+            None,
+            false,
+        );
```

Comment:
What if there's more than one state that can transition to `Final`? How is that handled?

E.g.:
```
ordered_elements: [
  symbol,
  { occurs: optional: type: text },
  { occurs: optional: type: number },
]
```